doc/developer/chips/10.rst

```diff
@@ -531,24 +531,76 @@ For more details on when the copy initializer would be called, please refer to
 Deinitialization
 ++++++++++++++++
 
-Records and classes can specify actions to take when a variable of that type is
-deinitialized. Record variables are deinitialized when they go out of scope and
-class variables are deinitialized when `delete` is called on them.
+Deinitialization is performed in these cases:
 
-To specify additional actions that occur upon deinitialization, specify a `proc
+ * Record variables are deinitialized when they go out of scope.
+ * Class variables are deinitialized when `delete` is called on them.
+
+To specify actions that occur upon deinitialization, user can specify a `proc
 deinit()` method. This method must have zero arguments and cannot be called
 directly. If a type does not specify a `proc deinit()` method, the compiler
 will generate one.
+It will be equivalent to a user-written `deinit` with an empty body.
+
+In addition to the actions specified by the `deinit` method, whether
+compiler-generated or user-defined, in the process of deinitialization:
 
-No matter whether the `deinit` method is compiler generated or user-defined,
-the compiler will ensure that in the process of deinitialization, all fields of
-record type are deinitialized. Additionally, the compiler will ensure that
-parent `deinit` methods are called in the case of inherited classes.
+ * parent `deinit` methods are called in the case of inherited classes, and
+ * all fields of record type are deinitialized.
 
 The expected use-case for the `deinit` method is to call `delete` on
 fields of class type, when appropriate. It is the responsibility of the
 class designer to `delete` fields when appropriate. The compiler will not
 automatically add `delete` calls for fields of class type.
+Conversely, the compiler adds `deinit` calls on fields of record type.
+These calls cannot be added by the user.
+
+Ensuring type safety
+********************
```

Comment:
Haven't gone through the PR yet, but I wanted to respond to Michael's comment first.

> I'm not sure we agreed that you could pass this to a function that never does dynamic dispatch on it.

I'm fairly certain we agreed that you could never pass this to a function during deinit.  I'm not sure at what point this became more specific.  The rationale still holds on this - if we were to allow operations on the "this" during any deinit, because we execute all deinits from child to base type, the object won't be in a valid state for any deinit method except for the first one that is called (because we decided that updating the type as we completed each deinitializer was probably a bad idea and at least unnecessary for now).